patches/0118-libavfilter-dnn-dnn_backend_torch-add-multi-device-s.patch

```diff
@@ -0,0 +1,301 @@
++    char **device_names;
++    int i = 0;
++    *nb_devices = 0;
++    while(device_str[i] != '\0') {
++        if(device_str[i] == '&')
++            *nb_devices += 1;
++        i++;
++    }
++    *nb_devices += 1;
++    device_names = (char **)av_mallocz(*nb_devices * sizeof(*device_names));
++    if (!device_names)
++        return NULL;
++
++    for (int i = 0; i < *nb_devices; i++) {
++        device_names[i] = av_strtok(device_str, "&", &saveptr);
++        device_str = NULL;
++    }
++    return device_names;
++}
++
+ static int extract_lltask_from_task(TaskItem *task, Queue *lltask_queue)
+ {
+     THModel *th_model = (THModel *)task->model;
+@@ -126,7 +153,7 @@ static int get_output_th(void *model, const char *input_name, int input_width, i
+     int ret = 0;
+     THModel *th_model = (THModel*) model;
+     THContext *ctx = &th_model->ctx;
+-    TaskItem task;
++    TaskItem task = { 0 };
+     THRequestItem *request;
+     DNNExecBaseParams exec_params = {
+         .input_name     = input_name,
+@@ -211,6 +238,7 @@ static DNNModel *dnn_load_model_th(const char *model_filename, DNNFunctionType f
+     THRequestItem *item = NULL;
+     THContext *ctx;
+     torch::jit::NameTensor first_param;
++    torch::jit::Module *jit_model;
+ 
+     model = (DNNModel *)av_mallocz(sizeof(DNNModel));
+     if (!model) {
+@@ -237,30 +265,43 @@ static DNNModel *dnn_load_model_th(const char *model_filename, DNNFunctionType f
+     else
+         torch::jit::setGraphExecutorOptimize(false);
+ 
+-    c10::Device device = c10::Device(ctx->options.device_name);
+-    if (device.is_cpu()) {
+-        ctx->options.device_type = torch::kCPU;
+-    } else if (device.is_xpu()) {
+-        if (!at::hasXPU()) {
+-            av_log(ctx, AV_LOG_ERROR, "No XPU device found\n");
+-            return NULL;
+-        }
+-        at::detail::getXPUHooks().initXPU();
+-        ctx->options.device_type = torch::kXPU;
+-    } else {
+-        av_log(ctx, AV_LOG_ERROR, "Not supported device:\"%s\"\n", ctx->options.device_name);
+-        goto fail;
++
++    th_model->device_names = th_separate_device_name(ctx->options.device_name, &th_model->nb_models);
++    if (!th_model->device_names) {
++        av_log(ctx, AV_LOG_ERROR, "could not parse devices names\n");
++        return NULL;
+     }
+ 
+-    try {
+-        th_model->jit_model = new torch::jit::Module;
+-        (*th_model->jit_model) = torch::jit::load(model_filename);
+-        th_model->jit_model->to(ctx->options.device_type);
+-    } catch (const c10::Error& e) {
+-        av_log(ctx, AV_LOG_ERROR, "Failed to load torch model\n");
++    th_model->jit_model_queue = ff_safe_queue_create();
++    if (!th_model->jit_model_queue) {
+         goto fail;
+     }
+-    first_param = *th_model->jit_model->named_parameters().begin();
++
++    for (int i = 0; i < th_model->nb_models; i++) {
++        c10::Device *device;
++        device = new c10::Device(th_model->device_names[i]);
++        if (device->is_xpu()) {
```

Comment:
check whether device is a NULL pointer firstly.

Reply:
done, Thanks.